### PR TITLE
[MODORDSTOR-354] - Implement business rules to update piece statues in batch job

### PR DESCRIPTION
**Pull request by Abdulkhakimov**
## Purpose
The PR aimed to implement pieces daily based batch job bussiness logic.

## Approach

Logic to proccess and update piece status was implemented using complex SQL statement which includes:

1. Update using CTE
2. Insert updated rows to audit outbox table

In the end will trigger audit outbox method which will send records to kafka.

## Learning
[MODORDSTOR-354](https://issues.folio.org/browse/MODORDSTOR-354)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Review by Dmitriy-Butramyou — APPROVED** [review]

File: src/main/java/org/folio/dao/PieceClaimingRepository.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        UPDATE %1$s.%2$s as pieces
+        SET jsonb = pieces.jsonb || jsonb_build_object(
+            'receivingStatus', 'Late',
+            'statusUpdatedDate', to_char(clock_timestamp(),'YYYY-MM-DD"T"HH24:MI:SS.MSTZH:TZM')::text,
+            'metadata', pieces.jsonb -> 'metadata' || jsonb_build_object(
+                'updatedDate', to_char(clock_timestamp(),'YYYY-MM-DD"T"HH24:MI:SS.MSTZH:TZM')::text,
+                'updatedByUserId', $1))
+        FROM %1$s.%3$s as titles
+        WHERE pieces.titleid = titles.id
+        AND (titles.jsonb ->> 'claimingActive')::boolean = TRUE
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dao/PieceClaimingRepository.java:29]: lets add index for field claimingActive, by default this flag would be disabled, so in this case with this index we can significanly reduce amount of records in WHERE cause

Comment (SerhiiNosko) [reply src/main/java/org/folio/dao/PieceClaimingRepository.java:29]: And also I think we can swap and make (titles.jsonb ->> 'claimingActive')::boolean = TRUE as first statement in WHERE cause. In this case our 'claimingActive' index will filter more records and joins for piece table where claimingActive=false will not occur

File: src/main/java/org/folio/services/piece/PieceClaimingService.java [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
+
+  public PieceClaimingService(PostgresClientFactory pgClientFactory, ExtensionRepository extensionRepository, PieceClaimingRepository pieceClaimingRepository) {
+    this.pgClientFactory = pgClientFactory;
+    this.extensionRepository = extensionRepository;
+    this.pieceClaimingRepository = pieceClaimingRepository;
+  }
+
+
+  public Future<Integer> processClaimedPieces(String tenantId) {
+    log.trace("processClaimedPieces, tenantId={}", tenantId);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/services/piece/PieceClaimingService.java:27]: Basically this job will be executed only once a day, so it's okay to have info level here- in this log

File: src/main/java/org/folio/services/piece/PieceClaimingService.java [hunk 3]
```diff
@@ -1,11 +1,33 @@
...
+  private final ExtensionRepository extensionRepository;
+  private final PieceClaimingRepository pieceClaimingRepository;
+
+  public PieceClaimingService(PostgresClientFactory pgClientFactory, ExtensionRepository extensionRepository, PieceClaimingRepository pieceClaimingRepository) {
+    this.pgClientFactory = pgClientFactory;
+    this.extensionRepository = extensionRepository;
+    this.pieceClaimingRepository = pieceClaimingRepository;
+  }
+
+
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply src/main/java/org/folio/services/piece/PieceClaimingService.java]: Thank you. Removed

File: src/main/java/org/folio/dao/PieceClaimingRepository.java (hunk 1 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/dao/PieceClaimingRepository.java]: Thanks. Improved

File: src/main/java/org/folio/services/piece/PieceClaimingService.java (hunk 2 above)
Comment (Abdulkhakimov) [reply src/main/java/org/folio/services/piece/PieceClaimingService.java]: Increased up to info level

**Pull request merged**